pkg/service/auth_context.go

```diff
@@ -63,53 +62,119 @@ func (authContext *AuthContext) evaluateAuthConfig(ctx context.Context, config c
 	}
 }
 
-func (authContext *AuthContext) evaluateAuthConfigs(authConfigs []common.AuthConfigEvaluator, cb configCallback) error {
-	errGroup, ctx := errgroup.WithContext(context.Background())
+func (authContext *AuthContext) evaluateOneAuthConfig(authConfigs []common.AuthConfigEvaluator, cb evaluateCallback) error {
+	ctx, cancel := context.WithCancel(context.Background())
+	waitGroup := new(sync.WaitGroup)
+	size := len(authConfigs)
+	errorChannel := make(chan error, size)
+	successChannel := make(chan bool, size)
+
+	waitGroup.Add(size)
+
+	go func() {
+		waitGroup.Wait()
+		close(errorChannel)
+		close(successChannel)
+	}()
 
 	for _, authConfig := range authConfigs {
 		objConfig := authConfig
-		errGroup.Go(func() error {
-			return authContext.evaluateAuthConfig(ctx, objConfig, cb)
-		})
+		go func() {
+			defer waitGroup.Done()
+
+			err := authContext.evaluateAuthConfig(ctx, objConfig, cb)
+			if err != nil {
+				errorChannel <- err
+			} else {
+				successChannel <- true
+				cancel() // cancels the context if at least one thread succeeds
+			}
+		}()
 	}
-	if err := errGroup.Wait(); err != nil {
-		return err
-	} else {
-		authCxtLog.Info("Successfully fetched all auth objects.", "authConfigs", authConfigs)
+
+	success := <-successChannel
+	err := <-errorChannel
+
+	if success {
 		return nil
+	} else {
+		return err
+	}
+}
+
+func (authContext *AuthContext) evaluateAllAuthConfigs(authConfigs []common.AuthConfigEvaluator, cb evaluateCallback) error {
+	ctx, cancel := context.WithCancel(context.Background())
+	waitGroup := new(sync.WaitGroup)
+	size := len(authConfigs)
+	errorChannel := make(chan error, size)
+
+	waitGroup.Add(size)
+
+	go func() {
+		waitGroup.Wait()
+		close(errorChannel)
+	}()
+
+	for _, authConfig := range authConfigs {
+		objConfig := authConfig
+		go func() {
+			defer waitGroup.Done()
+
+			err := authContext.evaluateAuthConfig(ctx, objConfig, cb)
+			if err != nil {
+				errorChannel <- err
+				cancel() // cancels the context if at least one thread fails
+			}
+		}()
+	}
+
+	err := <-errorChannel
+	return err
+}
+
+func (authContext *AuthContext) evaluateAnyAuthConfig(authConfigs []common.AuthConfigEvaluator, cb evaluateCallback) {
+	ctx := context.Background()
+	size := len(authConfigs)
+	waitGroup := new(sync.WaitGroup)
+	waitGroup.Add(size)
+
+	for _, authConfig := range authConfigs {
+		objConfig := authConfig
+		go func() {
+			defer waitGroup.Done()
+
+			_ = authContext.evaluateAuthConfig(ctx, objConfig, cb)
```

Comment:
evaluation success == (err == nil)
evaluation failed == (err != nil)

Shouldn't be different reporting? like :  
a) evaluation: true/false
b) err (if some unexpected error happens)

Reply:
I was talking about that with @didierofrivia this morning. Having a non-boolean response as such would be an alternative and a better fit to using https://pkg.go.dev/golang.org/x/sync@v0.0.0-20201207232520-09787c993a3a/errgroup again. We would need to change all evaluators though and give them this new semantics. I'd suggest doing it as an improvement/refactoring in a future iteration. Right now, this is more about fixing behaviour though.